File: 2025-05-MAY-GFG-POTD/May 26 - Insert in Sorted Circular Linked List/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def sortedInsert(self, head: 'Node', data: int) -> 'Node':
        new_node = Node(data)

        # Case 1: Empty list
        if not head:
            new_node.next = new_node
            return new_node

        current = head

        while True:
            # Case: Insert in between current and current.next
            if current.data <= data <= current.next.data:
                break

            # Case: At rotation point (max -> min)
            if current.data > current.next.data:
                if data >= current.data or data <= current.next.data:
                    break

            current = current.next

            # Completed full circle
            if current == head:
                break

        new_node.next = current.next
        current.next = new_node

        # Return new head if new node is the new smallest
        return head if head.data <= data else new_node

File: 2025-05-MAY-GFG-POTD/May 26 - Insert in Sorted Circular Linked List/Solution.cpp
/*
struct Node {
    int data;
    Node* next;

    Node(int x) {
        data = x;
        next = nullptr;
    }
};
*/

class Solution {
public:
    Node* sortedInsert(Node* head, int data) {
        Node* newNode = new Node(data);

        // Case 1: Empty list
        if (head == nullptr) {
            newNode->next = newNode;
            return newNode;
        }

        Node* current = head;

        while (true) {
            // Case: Insert in between current and next
            if (current->data <= data && data <= current->next->data) {
                break;
            }

            // Case: At rotation point (max -> min)
            if (current->data > current->next->data) {
                if (data >= current->data || data <= current->next->data) {
                    break;
                }
            }

            current = current->next;

            // Completed full circle
            if (current == head) {
                break;
            }
        }

        newNode->next = current->next;
        current->next = newNode;

        // Return new head if newNode is smaller than head
        return (head->data <= data) ? head : newNode;
    }
};
